### ISSUE TITLE
Remove kvdb WAL

### ISSUE THREAD

**Issue by yiwu-arbug** [tikv/tikv#9155]
## Development Task

TiKV essentially have two WAL, one as raft log stored in raftdb, one as rocksdb WAL. Theoretically we can get rid of the rocksdb WAL, but we currently make use of  it to store states like region split/merge, and issue fsync to make sure the state is persisted. The biggest benefit of removing rocksdb WAL is *not* to reduce IO, but to decouple apply thread from IO tasks.

While there's not immediate plan to remove the WAL, this issue keep track of discussions around the task.

**Comment by yiwu-arbug** [comment]
cc @Little-Wallace @BusyJay feel free to fill in more details of current discussion result here. Thanks.

**Comment by BusyJay** [comment]
If WAL is removed, does rocksdb still provide atomicity and order? For example, a write batch should not only lose some keys while save others. A write batch written after a lost one should also be lost.

**Comment by yiwu-arbug** [comment]
> If WAL is removed, does rocksdb still provide atomicity and order? For example, a write batch should not only lose some keys while save others. A write batch written after a lost one should also be lost.

Yes, but we need to enable the atomic flush feature, which may need some testing.

**Comment by BusyJay** [comment]
The major problems need to be solved are admin commands.

For admin changes that modify epoch like conf change/split/merge, a solution is recording the epoch of the region and its dependency in raft db. For example, if region a is split into several regions (b, c, a) at index i, then a record like (b, init epoch) -> (a, i) is recorded. If region b is merged into a at index j, then a record like (a, merged eoch) -> (b, j) should be recorded. In such way, when a region is recovered, it should not be started until its dependency has applied to the given index.

For log compaction, log should not be removed until flushed apply index is beyond the compact index.

For ingesting sst files, we need to wait until the apply index is flushed before deleting those files. Perhaps, we need to manually flush to avoid running out of space.

This should probably be proposed as an RFC.

**Comment by BusyJay** [comment]
If rocksdb still provide atomicity and order, we can still rely on apply index to tell if all previous logs are applied. Because order is guaranteed, so the following situations should not exist:
1. region a is revert back before splitting and the split region b's apply index is larger than initial index.
2. region b is revert back before merging and the target region a's apply index is larger than commit merge index.

**Comment by solotzg** [comment]
Data in raft engine is actually the WAL of apply processes. For tiflash, normal write raft cmd will be recorded in memory until admin cmd(like split/merge/compactlog) occurs. Then fsync memory cache of region into column storage engine and write region state and apply state to kv engine of proxy(a dynamic c lib extended from TiKV). When restarted, apply process could replay raft log from last applied index. Atomicity and idempotency of tiflash really depend on atomicity and order of storage engine of TiKV.

**Comment by yiwu-arbug** [comment]
> The major problems need to be solved are admin commands.
> 
> For admin changes that modify epoch like conf change/split/merge, a solution is recording the epoch of the region and its dependency in raft db. For example, if region a is split into several regions (b, c, a) at index i, then a record like (b, init epoch) -> (a, i) is recorded. If region b is merged into a at index j, then a record like (a, merged eoch) -> (b, j) should be recorded. In such way, when a region is recovered, it should not be started until its dependency has applied to the given index.
> 
> For log compaction, log should not be removed until flushed apply index is beyond the compact index.
> 
> For ingesting sst files, we need to wait until the apply index is flushed before deleting those files. Perhaps, we need to manually flush to avoid running out of space.
> 
> This should probably be proposed as an RFC.

For log compaction and file ingestion, we can probably make use of RocksDB's flush complete callback and the `ReadTier::read_tier` to advance the flushed apply index.

**Comment by yiwu-arbug** [comment]
> Data in raft engine is actually the WAL of apply processes. For tiflash, normal write raft cmd will be recorded in memory until admin cmd(like split/merge/compactlog) occurs. Then fsync memory cache of region into column storage engine and write region state and apply state to kv engine of proxy(a dynamic c lib extended from TiKV). When restarted, apply process could replay raft log from last applied index. Atomicity and idempotency of tiflash really depend on atomicity and order of storage engine of TiKV.

@solotzg thanks for chime in. Does TiFlash assume data is persisted once the apply state is written to kv engine?

**Comment by solotzg** [comment]
> > Data in raft engine is actually the WAL of apply processes. For tiflash, normal write raft cmd will be recorded in memory until admin cmd(like split/merge/compactlog) occurs. Then fsync memory cache of region into column storage engine and write region state and apply state to kv engine of proxy(a dynamic c lib extended from TiKV). When restarted, apply process could replay raft log from last applied index. Atomicity and idempotency of tiflash really depend on atomicity and order of storage engine of TiKV.
> 
> @solotzg thanks for chime in. Does TiFlash assume data is persisted once the apply state is written to kv engine?

@yiwu-arbug Yes. But not like TiKV, TiFlash only writes apply state to kv engine when applying admin command. Actually, for TiFlash, there is no data in write/default/lock cf in kv engine. Just like: https://github.com/solotzg/tikv/blob/tiflash-proxy-master/components/raftstore/src/store/fsm/apply.rs#L1375 https://github.com/solotzg/tikv/blob/tiflash-proxy-master/components/raftstore/src/store/fsm/apply.rs#L1183, we only put region meta and state info to write batch.
